### PR TITLE
feat: Validate field presence and key integrity for translations.txt

### PR DESCRIPTION
**Pull request by aababilov**
Checks that fields and values provided in translations are correct and checks that a
translation references existing entity.

This validator gracefully handles legacy Google translations format. If translations.txt
does not follow the new format, then no validations are performed.

### CODE REVIEW

File: core/src/main/java/org/mobilitydata/gtfsvalidator/validator/ValidatorLoader.java [hunk 1]
```diff
@@ -141,12 +141,18 @@ public ValidatorLoader(ImmutableList<String> validatorPackages) throws Validator
       throws ValidatorLoaderException {
     Constructor<T> constructor = chooseConstructor(validatorClass);
 
+    // Indicates that the full GtfsFeedContainer needs to be injected.
```
Comment (lionel-nj) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/validator/ValidatorLoader.java:144]: Could you explain why this change was needed please?

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerGenerator.java [hunk 2]
```diff
@@ -201,10 +203,58 @@ public TypeSpec generateGtfsContainerClass() {
...
+          .endControlFlow();
+      method.addStatement(
+          "return Optional.ofNullable($L.getOrDefault(keys[0], null))",
+          byKeyMapName(fileDescriptor.primaryKey().get().name()));
+    } else if (fileDescriptor.sequenceKey().isPresent() && fileDescriptor.firstKey().isPresent()) {
+      method
+          .beginControlFlow("if (keys.length != 2)")
+          .addStatement("return Optional.empty()")
+          .endControlFlow();
+      // TODO(aababilov): Implement this.
```
Comment (lionel-nj) [inline processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerGenerator.java:243]: Is it still something to do here?

**Review by lionel-nj — CHANGES_REQUESTED** [review]
Thanks @aababilov for this contribution! I like how this adapts to unofficial tables! You will find a few questions in-line. 
Could you also add documentation for the new classes/methods as well as unit test for `TranslationValidator` please? :)

**Comment by isabelle-dr** [comment]
Thank you for this contribution @aababilov!

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import org.mobilitydata.gtfsvalidator.notice.ValidationNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsFeedContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTranslation;
+import org.mobilitydata.gtfsvalidator.table.GtfsTranslationTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTranslationTableLoader;
+
+/**
+ * Validates that translations are provided in accordance with GTFS Specification. Supports two
+ * types of specifications * unofficial Google extension * official GTFS specification Checks that
```
Comment (barbeau) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java:34]: >unofficial Google extension

Is this extension documented somewhere that you could link to from here?

Also, it looks like the auto-formatting destroyed the bulleted list here - I'd suggest re-formatting in a way that keeps the list and complies with the Google Java Format.

Comment (barbeau) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java:34]: @lionel-nj As you work on profiles, perhaps pulling checks for the unofficial Google extension into a profile would be a good use case?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/validator/ValidatorLoader.java (hunk 1 above)
Comment (aababilov) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/validator/ValidatorLoader.java:144]: Please see usage example in TranslationsValidator.java. translations.table_name may reference any table, so we locate it by name in GtfsFeedContainer.

```
  @Inject
  TranslationsValidator(
      GtfsTranslationTableContainer translationTable, GtfsFeedContainer feedContainer) {
    this.translationTable = translationTable;
    this.feedContainer = feedContainer;
  }
  ...

    Optional<GtfsTableContainer<?>> table = feedContainer.getTable(translation.tableName());
    if (!table.isPresent()) {
      noticeContainer.addValidationNotice(new TranslationUnknownTableNameNotice(translation));
      return;
    }

```

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerGenerator.java (hunk 2 above)
Comment (aababilov) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerGenerator.java:243]: Yes, we need to implement this. One way to do it is to introduce a hash map<composite_key, entity>. The existing containers for stop_times and shapes do not support quick lookup by composite key - they just allow to retrieve a list of all entities that have the same first key (e.g., trip_id).

**Comment by aababilov** [comment]
> Thank you for this contribution @aababilov!

My pleasure!

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java (hunk 3 above)
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java:34]: Sorry for bad formatting!

Google translation.txt extension is hard to support in open-source since the "Google" translations.txt has different columns:

* `trans_id`
* `lang` (not `language`!)
* `translation`

Fields `table_name`, `field_name` and `language` are required in GTFS standard but not used in Google extension. So, it is not enough to add a plugin for validation of Google translations - instead, the schema has to be patched and officially required fields become optional in order to support Google translations.

**Comment by aababilov** [comment]
I am splitting this change into pieces. https://github.com/MobilityData/gtfs-validator/pull/979

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java (hunk 3 above)
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationsValidator.java]: Yes, that was a bad copypaste from the internal code. Fixed.

**Review by asvechnikov2 — APPROVED** [review]
Thank you for exporting this, looks much better with GtfsFeedContainer! LGTM!

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+   * @return whether the field is missing or unexpected
+   */
+  private static boolean isMissingOrUnexpectedField(
+      GtfsTranslation translation,
+      boolean actualPresence,
+      boolean expectedPresence,
+      String fieldName,
+      String fieldValue,
+      NoticeContainer noticeContainer) {
+    if (expectedPresence != actualPresence) {
```
[the remark this answers is not in the document]
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java]: Sure, good point!

**Comment by aababilov** [comment]
Thanks for review!
